app/controllers/users_controller.rb

```diff
@@ -89,4 +91,69 @@ def destroy
     flash[:notice] = "No user found with ID #{params[:id]}."
     redirect_to(users_path)
   end
+
+  def export_attendance_csv
+    @user_role = if session[:user_id]
+                   Customer.where(id: session[:user_id]).first.role
+                 else
+                   'not_logged_in'
+                 end
+    unless @user_role == 'admin'
+      flash[:notice] = 'You do not have admin permissions.'
+      return redirect_to(users_path)
+    end
+
+    # NOTE csv files are a bad idea for big data larger than this anyway, if the scale of this
+    # program is ever increased beyond this point, there needs to be some kind of filtering
+    users = Customer.order('last_name').take(10_000)
+    events = Event.order('date').take(10_000)
+
+    # create csv string
+    csv_string = CSV.generate do |csv|
+      # set the header row first
+      header = ['Last name', 'First name', 'total attendances']
+
+      # push a column for every meeting
+      event_index = header.length
+      # this hashmap is for keeping track of which column corresponds to an event id,
+      # because we have a manual query that would be clumsy to sort
+      id_map = {}
+      events.each do |event|
+        header.push(event.title)
+        id_map[event.id] = event_index
+        event_index += 1
+      end
+      csv << header
+
+      # fill out table rows according to every user
+      users.each do |user|
+        row = [user.last_name, user.first_name, '']
+        # start the row empty
+        (0..id_map.size).each do
+          row.push('')
+        end
+        # manual query instead of needing to have an existence query for every combination of
+        # customer and event
+        query = ActiveRecord::Base.connection.execute(
+          "SELECT customers_events.event_id FROM customers_events WHERE customers_events.customer_id = #{user.id};"
+        )
```

Comment:
We can take this as is for sprint 3, but we should definitely look into optimization in sprint 4, if not now.

Currently, this generates a separate sql query for every single user. We could probably make some JOIN query with the `customers_events` table so that we don't generate so much overhead during its construction.

Reply:
This is better than the original which ran a query for every cell in the table, but I guess we could go one step further in sprint 4.

Reply:
Now that I think about it, using one query would also prevent issues regarding database changes between multiple queries.

Reply:
Yeah, that's something I didn't think of. We can merge these changes for sprint 3 and tackle optimization in sprint 4.